platform/consul/controller.go

```diff
@@ -0,0 +1,171 @@
+	}
+
+	endpoints := c.getCatalogService(name, nil)
+	if len(endpoints) == 0 {
+		return nil, false
+	}
+
+	return convertService(endpoints), true
+}
+
+func (c *Controller) getServices() map[string][]string {
+	data, _, err := c.client.Catalog().Services(nil)
+	if err != nil {
+		glog.Warningf("Could not retrieve services from consul: %v", err)
+		return make(map[string][]string)
+	}
+
+	return data
+}
+
+func (c *Controller) getCatalogService(name string, q *api.QueryOptions) []*api.CatalogService {
+	endpoints, _, err := c.client.Catalog().Service(name, "", q)
+	if err != nil {
+		glog.Warningf("Could not retrieve service catalogue from consul: %v", err)
+		return []*api.CatalogService{}
+	}
+
+	return endpoints
+}
+
+// Instances retrieves instances for a service and its ports that match
+// any of the supplied tags. All instances match an empty tag list.
+func (c *Controller) Instances(hostname string, ports []string, tags model.TagsList) []*model.ServiceInstance {
+	// Get actual service by name
+	name, err := parseHostname(hostname)
+	if err != nil {
+		glog.V(2).Infof("parseHostname(%s) => error %v", hostname, err)
+		return nil
+	}
+
+	endpoints := c.getCatalogService(name, nil)
+
+	instances := []*model.ServiceInstance{}
+	for _, endpoint := range endpoints {
+		instance := convertInstance(endpoint)
+		if tags.HasSubsetOf(instance.Tags) && portMatch(instance, ports) {
+			instances = append(instances, instance)
+		}
+	}
+
+	return instances
+}
+
+// returns true if an instance's port matches with any in the provided list
+func portMatch(instance *model.ServiceInstance, ports []string) bool {
+	if len(ports) == 0 {
+		return true
+	}
+
+	for _, port := range ports {
+		if instance.Endpoint.ServicePort.Name == port {
+			return true
+		}
+	}
+
+	return false
+}
+
+// HostInstances lists service instances for a given set of IPv4 addresses.
+func (c *Controller) HostInstances(addrs map[string]bool) []*model.ServiceInstance {
+	data := c.getServices()
+	out := make([]*model.ServiceInstance, 0)
+	for svcName := range data {
+		endpoints := c.getCatalogService(svcName, nil)
+		for _, endpoint := range endpoints {
+			if addrs[endpoint.ServiceAddress] {
+				out = append(out, convertInstance(endpoint))
+			}
+		}
+	}
+
+	return out
+}
+
+// Run all controllers until a signal is received
+func (c *Controller) Run(stop <-chan struct{}) {
+	c.monitor.Start(stop)
+}
+
+// AppendServiceHandler implements a service catalog operation
+func (c *Controller) AppendServiceHandler(f func(*model.Service, model.Event)) error {
+	c.monitor.AppendServiceHandler(func(instances []*api.CatalogService, event model.Event) error {
+		f(convertService(instances), event)
+		return nil
+	})
+	return nil
+}
+
+// AppendInstanceHandler implements a service catalog operation
+func (c *Controller) AppendInstanceHandler(f func(*model.ServiceInstance, model.Event)) error {
+	c.monitor.AppendInstanceHandler(func(instance *api.CatalogService, event model.Event) error {
```

Comment:
consul has no formal `Service` object and `CatalogService` is the representation of a service instance.  I went with `instance` to avoid confusion with model.Service and model.ServiceInstance